lib/Graph/TensorLayout.cpp

```diff
@@ -0,0 +1,277 @@
+};
+
+static TensorLayoutDescription layoutNHWC(dimsNHWC);
+static TensorLayoutDescription layoutNCHW(dimsNCHW);
+static TensorLayoutDescription layoutHWNC(dimsHWNC);
+static TensorLayoutDescription layout0D({});
+static TensorLayoutDescription layout1D(dims1D);
+static TensorLayoutDescription layout2D(dims2D);
+static TensorLayoutDescription layout3D(dims3D);
+static TensorLayoutDescription layout4D(dims4D);
+static TensorLayoutDescription layout5D(dims5D);
+static TensorLayoutDescription layout6D(dims6D);
+
+/// Glow layouts for any specific number of dimensions.
+static std::array<TensorLayoutDescription, max_tensor_dimensions + 1>
+    layoutsForDims = {
+        layout0D, layout1D, layout2D, layout3D, layout4D, layout5D, layout6D,
+};
+
+static std::unordered_map<std::string, TensorLayoutDescription *>
+    layoutNameToLayoutDescription = {{
+        {"NCHW", &layoutNCHW},
+        {GLOW_DEFAULT_4D_LAYOUT, &layoutNHWC},
+        {"HWNC", &layoutHWNC},
+        {"N", &layout1D},
+    }};
+
+const std::array<TensorLayoutDescription, max_tensor_dimensions + 1> &
+TensorLayoutCommon::getLayoutsForDims() const {
+  return layoutsForDims;
+}
+
+static const TensorLayoutDescription *
+getLayoutFromName(const std::string &name) {
+  auto it = layoutNameToLayoutDescription.find(name);
+  if (it != layoutNameToLayoutDescription.end()) {
+    return it->second;
+  }
+  return nullptr;
+}
+
+const TensorLayoutDescription
+TensorLayoutCommon::getDefaultNDLayout(unsigned dims) const {
+  DCHECK_LE(dims, max_tensor_dimensions) << "Too many dimensions";
+  return getLayoutsForDims()[dims];
+}
+
+const TensorLayoutDescription
+TensorLayoutCommon::getNthInputLayoutRequirements(const Node *node,
+                                                  size_t n) const {
+  DCHECK_LT(n, node->getNumInputs()) << "Wrong input number";
+  auto dims = node->getNthInput(n).getType()->dims();
+  DCHECK_LE(dims.size(), max_tensor_dimensions) << "Too many dimensions";
+  if (const auto *TN = llvm::dyn_cast<TransposeNode>(node)) {
+    // The layout for the input of transpose is the same as the layout of the
+    // operation's result producing this input.
+    auto input = TN->getInput();
+    return getNthResultLayoutRequirements(input.getNode(), input.getResNo());
+  }
+  if (const auto *QN = llvm::dyn_cast<QuantizeNode>(node)) {
+    auto input = QN->getInput();
+    return getNthResultLayoutRequirements(input.getNode(), input.getResNo());
+  }
+  if (const auto *QPN = llvm::dyn_cast<QuantizationProfileNode>(node)) {
+    switch (n) {
+    case QuantizationProfileNode::InputIndices::InputIdx: {
+      auto input = QPN->getInput();
+      return getNthResultLayoutRequirements(input.getNode(), input.getResNo());
+    }
+    default:
+      return getLayoutsForDims()[dims.size()];
+    }
+  }
+  return getLayoutsForDims()[dims.size()];
+}
+
+/// \returns The index of node \p N input \p IN. NumInputs if not found.
+static unsigned getInputIdx(const Node *N, NodeValue IN) {
```

Comment:
Why is IN upper-cased? I don't think we often use this kind of naming for NodeValues.